src/Init/Data/List/Control.lean

```diff
@@ -40,24 +40,54 @@ Finally, we rarely use `mapM` with something that is not a `Monad`.
 Users that want to use `mapM` with `Applicative` should use `mapA` instead.
 -/
 
+/--
+Applies the monadic action `f` on every element in the list, left-to-right, and returns the list of
+results.
+
+See `List.forM` for the variant that discards the results.
+See `List.mapA` for the variant that works with `Applicative`.
+-/
 @[inline]
 def mapM {m : Type u → Type v} [Monad m] {α : Type w} {β : Type u} (f : α → m β) (as : List α) : m (List β) :=
   let rec @[specialize] loop
     | [],      bs => pure bs.reverse
     | a :: as, bs => do loop as ((← f a)::bs)
   loop as []
 
+/--
+Applies the applicative action `f` on every element in the list, left-to-right, and returns the list of
+results.
+
+NB: If `m` is also a `Monad`, then using `mapM` can be more efficient.
+
+See `List.forA` for the variant that discards the results.
+See `List.mapM` for the variant that works with `Monad`.
+-/
```

Comment:
```suggestion

**Warning**: this function is not tail-recursive, meaning that it may fail with stack overflow on long lists.
-/
```

Reply:
Is this related to the
```
NB: If `m` is also a `Monad`, then using `mapM` can be more efficient.
```
comment or independent?

Reply:
Independent, although I'm not sure it's even possible to write this in a tail-recursive way. It's much easier to get a `bind` chain to have the desired form for tail recursion than a `seq` chain.

Reply:
Thanks!